File: app/src/main/java/ua/fvadevand/reminderstatusbar/ui/ReminderEditFragment.java
package ua.fvadevand.reminderstatusbar.ui;

import android.arch.lifecycle.LiveData;
import android.arch.lifecycle.ViewModelProviders;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.CheckBox;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.TextView;

import java.util.Calendar;

import ua.fvadevand.reminderstatusbar.Const;
import ua.fvadevand.reminderstatusbar.R;
import ua.fvadevand.reminderstatusbar.adapters.IconAdapter;
import ua.fvadevand.reminderstatusbar.data.models.Reminder;
import ua.fvadevand.reminderstatusbar.dialogs.AlarmSetDialog;
import ua.fvadevand.reminderstatusbar.dialogs.AlarmSetDialog.OnAlarmSetListener;
import ua.fvadevand.reminderstatusbar.dialogs.IconsDialog;
import ua.fvadevand.reminderstatusbar.listeners.FabVisibilityChangeListener;
import ua.fvadevand.reminderstatusbar.utilities.IconUtils;
import ua.fvadevand.reminderstatusbar.utilities.ReminderDateUtils;

public class ReminderEditFragment extends Fragment
        implements View.OnClickListener, OnAlarmSetListener,
        IconAdapter.OnIconClickListener {

    public static final String TAG = "ReminderEditFragment";
    private static final String ARG_REMINDER_ID = "reminder_id";
    private FabVisibilityChangeListener mFabVisibilityChangeListener;
    private boolean isEditMode;
    private EditText mTitleView;
    private EditText mTextView;
    private ImageButton mIconBtn;
    private Button mNotifyBtn;
    private TextView mTimeView;
    private CheckBox mDelayNotificationView;
    private Calendar mCalendar;
    private int mIconResId;
    private RemindersViewModel mViewModel;
    private LiveData<Reminder> mCurrentReminderLive;
    private long mCurrentReminderId;

    public ReminderEditFragment() {
    }

    public static ReminderEditFragment newInstance(long reminderId) {
        ReminderEditFragment fragment = new ReminderEditFragment();
        Bundle args = new Bundle();
        args.putLong(ARG_REMINDER_ID, reminderId);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            mCurrentReminderId = getArguments().getLong(ARG_REMINDER_ID);
            isEditMode = mCurrentReminderId != Const.NEW_REMINDER_ID;
        } else {
            mCurrentReminderId = Const.NEW_REMINDER_ID;
        }
        mViewModel = ViewModelProviders.of(getActivity()).get(RemindersViewModel.class);
        if (isEditMode) {
            mCurrentReminderLive = mViewModel.getReminderById(mCurrentReminderId);
        }
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_reminder_edit, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initView(view);
        if (isEditMode) {
            mCurrentReminderLive.observe(this, reminder -> {
                mCurrentReminderLive.removeObservers(ReminderEditFragment.this);
                fillView(reminder);
            });
        } else {
            setVisibilityTimeView(false);
        }
    }

    private void initView(View view) {
        mTitleView = view.findViewById(R.id.et_edit_reminder_title);
        mTitleView.requestFocus();
        mTextView = view.findViewById(R.id.et_edit_reminder_text);
        (mIconBtn = view.findViewById(R.id.btn_edit_reminder_icon)).setOnClickListener(this);
        mIconResId = R.drawable.ic_notif_edit;
        mIconBtn.setImageResource(mIconResId);
        view.findViewById(R.id.btn_edit_reminder_time).setOnClickListener(this);
        (mNotifyBtn = view.findViewById(R.id.btn_edit_reminder_notify)).setOnClickListener(this);
        mTimeView = view.findViewById(R.id.tv_edit_reminder_time);
        mDelayNotificationView = view.findViewById(R.id.cb_edit_reminder_delay_notification);
        mCalendar = Calendar.getInstance();
    }

    private void fillView(@Nullable Reminder reminder) {
        if (reminder == null) {
            clearView();
            return;
        }
        mTitleView.setText(reminder.getTitle());
        mTextView.setText(reminder.getText());
        if (getContext() != null) {
            mIconResId = IconUtils.getIconResId(getContext(), reminder.getIconName());
            mIconBtn.setImageResource(mIconResId);
        }
        if (reminder.getTimestamp() > System.currentTimeMillis()) {
            setVisibilityTimeView(true);
            mCalendar.setTimeInMillis(reminder.getTimestamp());
            mTimeView.setText(getNotificationTimeString(mCalendar));
        } else {
            setVisibilityTimeView(false);
        }
    }

    private void clearView() {
        mTitleView.getText().clear();
        mTitleView.requestFocus();
        mTextView.getText().clear();
        setVisibilityTimeView(false);
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.btn_edit_reminder_icon:
                showIconsDialog();
                break;
            case R.id.btn_edit_reminder_time:
                showSetAlarmDialog();
                break;
            case R.id.btn_edit_reminder_notify:
                saveAndNotifyReminder();
                break;
        }
    }

    private void showSetAlarmDialog() {
        AlarmSetDialog.newInstance(mCalendar).show(getChildFragmentManager(), AlarmSetDialog.TAG);
    }

    private void showIconsDialog() {
        new IconsDialog().show(getChildFragmentManager(), IconsDialog.TAG);
    }

    private void saveAndNotifyReminder() {
        String title = mTitleView.getText().toString().trim();
        if (TextUtils.isEmpty(title)) {
            mTitleView.setError(getString(R.string.reminder_edit_error_empty_title));
            return;
        }
        String text = mTextView.getText().toString().trim();
        long timeInMillis;
        if (mDelayNotificationView.isChecked()) {
            timeInMillis = mCalendar.getTimeInMillis();
        } else {
            timeInMillis = System.currentTimeMillis();
        }
        Reminder reminder = new Reminder();
        if (isEditMode) {
            reminder.setId(mCurrentReminderId);
        }
        reminder.setTitle(title);
        reminder.setTimestamp(timeInMillis);
        reminder.setText(text);
        reminder.setIconName(IconUtils.getIconName(getContext(), mIconResId));
        mViewModel.insertReminder(reminder);
        isEditMode = false;
        clearView();
    }

    @Override
    public void onAlarmSet(Calendar calendar) {
        mCalendar = calendar;
        setVisibilityTimeView(true);
        mTimeView.setText(getNotificationTimeString(mCalendar));
        mNotifyBtn.setText(R.string.edit_reminder_action_save);
    }

    private String getNotificationTimeString(Calendar calendar) {
        return getString(R.string.edit_reminder_notification_time, ReminderDateUtils.getNotificationTime(getContext(), calendar));
    }

    private void setVisibilityTimeView(boolean isVisibility) {
        int visibility;
        if (isVisibility) {
            visibility = View.VISIBLE;
        } else {
            visibility = View.INVISIBLE;
        }
        mDelayNotificationView.setVisibility(visibility);
        mDelayNotificationView.setChecked(isVisibility);
        mTimeView.setVisibility(visibility);
    }

    @Override
    public void onIconClick(int iconResId) {
        mIconResId = iconResId;
        mIconBtn.setImageResource(mIconResId);
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof FabVisibilityChangeListener) {
            mFabVisibilityChangeListener = (FabVisibilityChangeListener) context;
        } else {
            throw new RuntimeException(context.toString()
                    + " must implement FabVisibilityChangeListener");
        }
    }

    @Override
    public void onStart() {
        super.onStart();
        mFabVisibilityChangeListener.fabVisibilityChange(false);
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mFabVisibilityChangeListener = null;
    }
}

File: app/src/main/java/ua/fvadevand/reminderstatusbar/ui/ReminderEditFragment.kt
package ua.fvadevand.reminderstatusbar.ui

import android.animation.ValueAnimator
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.WindowManager
import android.view.inputmethod.EditorInfo
import android.widget.EditText
import androidx.core.animation.doOnEnd
import androidx.core.view.ViewCompat
import androidx.core.view.isVisible
import androidx.core.widget.addTextChangedListener
import androidx.fragment.app.activityViewModels
import com.google.android.material.bottomsheet.BottomSheetBehavior
import ua.fvadevand.reminderstatusbar.Const
import ua.fvadevand.reminderstatusbar.R
import ua.fvadevand.reminderstatusbar.data.models.PeriodType
import ua.fvadevand.reminderstatusbar.data.models.PeriodType.PeriodTypes
import ua.fvadevand.reminderstatusbar.data.models.Reminder
import ua.fvadevand.reminderstatusbar.data.models.ReminderStatus
import ua.fvadevand.reminderstatusbar.databinding.FragmentReminderEditBinding
import ua.fvadevand.reminderstatusbar.ui.dialogs.AlarmSetDialog
import ua.fvadevand.reminderstatusbar.ui.dialogs.AlarmSetDialog.OnAlarmSetListener
import ua.fvadevand.reminderstatusbar.ui.dialogs.IconsDialog
import ua.fvadevand.reminderstatusbar.utils.getNotificationTime
import ua.fvadevand.reminderstatusbar.utils.hideSoftKeyboard
import ua.fvadevand.reminderstatusbar.utils.setImageResourceName
import ua.fvadevand.reminderstatusbar.utils.showSoftKeyboard
import ua.fvadevand.reminderstatusbar.utils.toResId
import ua.fvadevand.reminderstatusbar.utils.toResName
import ua.fvadevand.reminderstatusbar.utils.updateSystemWindowInsets
import java.util.Locale

class ReminderEditFragment : BaseBottomSheetDialogFragment(R.layout.fragment_reminder_edit),
    View.OnClickListener,
    OnAlarmSetListener,
    IconsDialog.OnIconClickListener {

    private val binding by fragmentProperty.bindingByView(
        FragmentReminderEditBinding::bind
    )
    private val viewModel: RemindersViewModel by activityViewModels()
    private var startTimeInMillis = System.currentTimeMillis()
    private var editMode = false
    private var iconResId = R.drawable.ic_notif_edit

    private var currentFocus: View? by fragmentProperty.delegateViewLifecycle()
    private var hasKeyboard = false

    @PeriodTypes
    private var periodType = PeriodType.ONE_TIME
    private var navBarHeight = -1
    private var behavior: BottomSheetBehavior<View>? by fragmentProperty.delegateViewLifecycle()
    private val currentReminderId by lazy {
        arguments?.getLong(ARG_REMINDER_ID) ?: Const.NEW_REMINDER_ID
    }

    companion object {
        const val TAG = "ReminderEditFragment"
        private const val DELAY_UP_DIALOG = 100L
        private const val DELAY_DOWN_DIALOG = 50L
        private const val ARG_REMINDER_ID = "REMINDER_ID"

        fun newInstance(reminderId: Long) = ReminderEditFragment().apply {
            arguments = Bundle().apply { putLong(ARG_REMINDER_ID, reminderId) }
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        dialog?.window?.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_RESIZE)
        activity?.window?.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_NOTHING)
        return super.onCreateView(inflater, container, savedInstanceState)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        dialog?.setOnShowListener {
            (view.parent as? View)?.let {
                behavior = BottomSheetBehavior.from(it).apply {
                    peekHeight = 0
                    skipCollapsed = true
                }
            }
        }
        editMode = currentReminderId != Const.NEW_REMINDER_ID
        initView()
        if (editMode) {
            viewModel.getReminder(currentReminderId) {
                it?.let { populateData(it) }
            }
        }
        dialog?.window?.decorView?.also { decorView ->
            ViewCompat.setOnApplyWindowInsetsListener(decorView) { v, insets ->
                if (navBarHeight == -1) navBarHeight = insets.systemWindowInsetBottom
                val heightDifference = insets.systemWindowInsetBottom - navBarHeight
                if (view.paddingBottom != heightDifference) {
                    val animator = ValueAnimator.ofInt(view.paddingBottom, heightDifference)
                    val duration = if (heightDifference == 0) DELAY_DOWN_DIALOG else DELAY_UP_DIALOG
                    animator.addUpdateListener { view.setPadding(0, 0, 0, it.animatedValue as Int) }
                    animator.doOnEnd { behavior?.state = BottomSheetBehavior.STATE_EXPANDED }
                    animator.duration = duration
                    animator.start()
                }

                ViewCompat.onApplyWindowInsets(
                    v,
                    insets.updateSystemWindowInsets(bottom = navBarHeight)
                )
            }
        }
    }

    override fun onIconClick(iconId: Int) {
        iconResId = iconId
        binding.btnEditReminderIcon.setImageResource(iconResId)
    }

    override fun onDestroyView() {
        activity?.window?.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_RESIZE)
        super.onDestroyView()
    }

    private fun initView() {
        binding.apply {
            val editTextFocusChangeListener = View.OnFocusChangeListener { v, hasFocus ->
                if (hasFocus) currentFocus = v
            }
            etEditReminderTitle.onFocusChangeListener = editTextFocusChangeListener
            etEditReminderText.onFocusChangeListener = editTextFocusChangeListener
            etEditReminderTitle.showSoftKeyboard()
            etEditReminderText.setHorizontallyScrolling(false)
            etEditReminderText.maxLines = 20
            etEditReminderText.setOnEditorActionListener { _, actionId, _ ->
                when (actionId) {
                    EditorInfo.IME_ACTION_DONE -> {
                        if (etEditReminderTitle.text.isNullOrEmpty()) {
                            etEditReminderTitle.requestFocus()
                        } else {
                            saveAndNotifyReminder()
                        }
                        true
                    }
                    else -> false
                }
            }
            btnEditReminderNotify.isEnabled = false
            etEditReminderTitle.addTextChangedListener(afterTextChanged = {
                btnEditReminderNotify.isEnabled = it.toString().trim().isNotEmpty()
            })
            btnEditReminderIcon.setOnClickListener(this@ReminderEditFragment)
            btnEditReminderIcon.setImageResource(R.drawable.ic_grid)
            btnEditReminderTime.setOnClickListener(this@ReminderEditFragment)
            btnEditReminderNotify.setOnClickListener(this@ReminderEditFragment)
            chipEditReminderRepeat.setOnCloseIconClickListener {
                it.isVisible = false
                periodType = PeriodType.ONE_TIME
            }
            chipEditReminderTime.setOnCloseIconClickListener {
                it.isVisible = false
                startTimeInMillis = System.currentTimeMillis()
                setActionText()
            }
            startTimeInMillis = System.currentTimeMillis()
        }
    }

    private fun populateData(reminder: Reminder) {
        binding.etEditReminderTitle.setText(reminder.title)
        binding.etEditReminderText.setText(reminder.text)
        binding.btnEditReminderIcon.run {
            iconResId = context.toResId(reminder.iconName)
            setImageResource(iconResId)
            setImageResourceName(reminder.iconName)
        }
        val now = System.currentTimeMillis()
        startTimeInMillis =
            if (reminder.status == ReminderStatus.PAUSED && now > reminder.timestamp) {
                PeriodType.getNextAlarmTimeByType(reminder.periodType, reminder.timestamp)
            } else {
                reminder.timestamp
            }
        if (startTimeInMillis > now) {
            binding.chipEditReminderTime.isVisible = true
            binding.chipEditReminderTime.text = context.getNotificationTime(startTimeInMillis)
        } else {
            binding.chipEditReminderTime.isVisible = false
        }
        periodType = reminder.periodType
        binding.chipEditReminderRepeat.isVisible = if (periodType == PeriodType.ONE_TIME) {
            false
        } else {
            binding.chipEditReminderRepeat.text = getRepeatString()
            true
        }
        setActionText()
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.btn_edit_reminder_icon -> showIconsDialog()
            R.id.btn_edit_reminder_time -> showSetAlarmDialog()
            R.id.btn_edit_reminder_notify -> saveAndNotifyReminder()
        }
    }

    private fun showSetAlarmDialog() {
        hasKeyboard = binding.root.paddingBottom != 0
        currentFocus?.hideSoftKeyboard()
        AlarmSetDialog.newInstance(startTimeInMillis, periodType)
            .show(childFragmentManager, AlarmSetDialog.TAG)
    }

    private fun showIconsDialog() {
        IconsDialog.newInstance().show(childFragmentManager, IconsDialog.TAG)
    }

    private fun saveAndNotifyReminder() {
        val title = binding.etEditReminderTitle.text.toString().trim()
        if (title.isEmpty()) {
            binding.etEditReminderTitle.error =
                getString(R.string.edit_reminder_error_empty_title)
            return
        }
        val text = binding.etEditReminderText.text.toString().trim()
        val status = when {
            periodType > PeriodType.ONE_TIME -> ReminderStatus.PERIODIC
            startTimeInMillis > System.currentTimeMillis() -> ReminderStatus.DELAYED
            else -> ReminderStatus.NOTIFYING
        }
        val reminder = Reminder(
            title,
            text,
            requireContext().toResName(iconResId),
            startTimeInMillis,
            status,
            periodType
        )
        if (editMode) {
            reminder.id = currentReminderId
        }
        viewModel.addReminder(reminder)
        editMode = false
        dismiss()
    }

    override fun onAlarmSet(alarmTimeInMillis: Long, @PeriodTypes periodType: Int) {
        restoreKeyboard()
        startTimeInMillis = alarmTimeInMillis
        binding.chipEditReminderTime.isVisible = true
        binding.chipEditReminderTime.text = context.getNotificationTime(startTimeInMillis)
        setActionText()
        this.periodType = periodType
        if (periodType > PeriodType.ONE_TIME) {
            binding.chipEditReminderRepeat.isVisible = true
            binding.chipEditReminderRepeat.text = getRepeatString()
        }
    }

    override fun onAlarmCancelled() {
        restoreKeyboard()
    }

    private fun restoreKeyboard() {
        if (hasKeyboard) {
            currentFocus?.run {
                showSoftKeyboard()
                if (this is EditText) setSelection(text.length)
            }
        }
    }

    private fun getRepeatString(): String {
        return getString(
            R.string.edit_reminder_repeat,
            getString(PeriodType.getPeriodTypeStringResId(periodType)).toLowerCase(Locale.getDefault())
        )
    }

    private fun setActionText() {
        val textResId =
            if (startTimeInMillis > System.currentTimeMillis()) R.string.edit_reminder_action_save
            else R.string.edit_reminder_action_notify
        binding.btnEditReminderNotify.setText(textResId)
    }

}
